// chromatic/chromatic-template.js
function template( arr, settings ) {
	var result = "";
	var templateFilename = "chromatic-template.ly";
	var r = String.raw;

	// console.error( 'settings', settings );

	if ( ! settings ) 
		settings = {
			tempo : 130,
		};

	// console.error( 'default settings', settings );
	function js2scm( k ) {
		k = k.replace( /([A-Z][a-z0-9]*)/g, (s0,s1)=>"-" + s1.toLowerCase() );
		k = k.replace( /^-/, '' );
		return k;
	}
	function value_js2scm( value ) {
		var s = '';
		for ( var i in value ) {
			var matched = /^(scm)(.*)$/.exec( i );
			if ( matched ) {
				switch ( typeof value[i] ) {
					case "string" :
					case "number" :
						s += `( ${js2scm(matched[2])} . ${value[i]} )`;
						break;
					case "boolean" :
						s += `( ${js2scm(matched[2])} . ${value[i] ? '#t' : '#f' } )`;
						break;
					// case "object" :
					// 	if ( Array.isArray( value[i] ) ) {
					// 		var arr=  value[i];
					// 		s += "( " ;
					// 		for ( var idx=0 ; idx<arr.length; idx++ ) {
					// 			s += 
					// 		}
					// 		s += ") " ;
					// 	}
					// 	break;
					default:
				}
			}
		}
		return  s;
	}

	function getSetting() {
		// `'(${s})`;
		return "'(" + value_js2scm( settings ) + ")" ;
	}

result += `
\\version "2.18.2"
\\include "aaron.ly"
\\include "${templateFilename}"
\\include "lilypond-book-preamble.ly"
\\language "aaron"

`;

	function createBlock( caption, comment, notes, lyrics ) {
		var result =  "";
		if ( caption ) {
			result +=
(`
\\markup \\bold \\italic {
    ${caption}
}
`);
		}

		if ( comment ) {
			let es = comment.split( /\n/ );
			for ( let i=0; i<es.length; i++ ) {
				result += "    %" + es[i] + "\n";
			}
		}

		switch ( settings.formatType  ) {
			case 'entire-diagram':
				result += ( r`
					#(define override-default ` + getSetting() + r` )
					\markup 
					  {
						#(create-markup-of-entire-scale-diagram #{ ` + notes  + r` #}
							(ly:assoc-get 'root-offset override-default 0 ) override-default )
					  }
					  #` + getSetting() 
					).replace( /^\s{5}/gm, '' );
				break;
			case 'diagram':
				result += ( r`
					#(define override-default ` + getSetting() + r` )
					#(define scale-diagram-def
						(create-markup-of-scale-diagram 
							#{ \absolute { ` + notes  + r` } #}
							(ly:assoc-get 'fret-positions override-default '(6 6 6 5 5 5 4 4 4 3 3 3 2 2 2 1 1 1) ) 
							(ly:assoc-get 'skip-count     override-default 0 ) 
							(ly:assoc-get 'root-offset    override-default 0 ) 
							override-default ))
					\markup {
						#(cdr (assq 'markup scale-diagram-def))
					}
					\makescore
						\absolute {
							#(cdr (assq 'accepted-notes scale-diagram-def))
						}
						#` + getSetting() + `
					`).replace( /^\s{5}/gm, '' );
				break;

			case 'absolute':
				result += ( r`\
					\makescore 
					  {
						` + notes + `
					  }
					  #` + getSetting() + `
					`).replace( /^\s{5}/gm, '' );
				break;
			case 'relative':
			default : 
				result += (r`\
					\makescore 
					  \relative do' {
						` + notes + r`
					  }
					  #` + getSetting() + r`
					`).replace( /^\s{5}/gm, '' );
				break;
		}
		return result;
	}

	for ( var i=0; i<arr.length; i++ ) {
		var e= arr[i];
		result += createBlock( e.caption, e.comment, e.notes, e.lyrics );
	}
	return result.trim();
}


	/////////////////////////////////////////////
	//
	//
	//
	/////////////////////////////////////////////

	function applyTemplate( noteArrays, settings ) {
		if ( Array.isArray( noteArrays ) ) {
		} else {
			noteArrays = [ noteArrays ] ;
		}

		var result = [];
		for ( var i=0; i< noteArrays.length; i++ ) {
			result[i] = {
				notes : toNotes(noteArrays[i]),
			};
		}
		return template( result, settings );
	};

if ( module && module.exports ) {
	module.exports.template = template;
	module.exports.applyTemplate = applyTemplate;
} else {
	let content = '';
	process.stdin.resume();
	process.stdin.on('data', function(buf) { content += buf.toString(); });
	process.stdin.on('end', function() {
		console.error( content );
		process.stdout.write( template( new Function( content ).apply() ) );
		process.stdout.write( "\n" );
	});
}
