// chromatic/settings.js
// Default values for the module 'chromatic'.
module.exports = {
		/*
		 * The default tempo for festival voice generator. 
		 * - (Fri, 04 May 2018 18:38:30 +0900) Added prefix scm
		 */
		scmFestivalTempo : 150,

		/*
		 * The default voice name for festival voice generator.
		 * History
		 * - (Fri, 04 May 2018 18:38:30 +0900) Added prefix scm
		 */
		scmFestivalVoice : 'voice_us2_mbrola',

		/*
		 * Default voice string filter for festival voice generator.
		 * History
		 * - (Fri, 04 May 2018 18:38:30 +0900) Renamed from voiceFilter() to festivalFilter().
		 */
		festivalFilter :(type, s)=>s,

		/*
		 *  "eventhandler"
		 *  "eventhandler" is used to filter various values which are generated
		 *  druing a writeScore() session. 
		 * 
		 *  Generally `filter` should not be an arrow function since it receives
		 *  ScaleGenerator object as `this` pointer.  But as a default value, it
		 *  is not necessary to receive this so it can be an arrow function.
		 * 
		 *  History
		 *  - (Sat, 28 Apr 2018 23:35:23 +0900) Created
		 *  - (Fri, 04 May 2018 17:21:44 +0900) Moved from ScaleGenerator to settings.js
		 *                                      Renamed from filter to eventhandler
		 *
		 */
		eventhandler : (eventtype,id,notes,text)=>text,

		/*
		 * The default value of a size which is refered by writeScore() function.
		 * This is applied in \includegraphics command of tex output file. 1.0
		 * means 100% width of its container document.
		 *
		 * History
		 * - (Fri, 04 May 2018 17:59:09 +0900) Renamed from size to texGraphicWidth.
		 */
		texGraphicWidth : '1.0',

		/*
		 * The default value of a size which is refered by writeScore() function.
		 */
		lyTextBefore : null,

		/*
		 * The default value of a size which is refered by writeScore()
		 * function.  If textAfter is an array, then it will be converted to a
		 * lilypond code block that specify a paper dimension.  In such case,
		 * it is interpreted as [h,w]. The unit is(maybe) inch.
		 */
		lyTextAfter : [5,3],
	};
